### PR TITLE
Check the random number distibution

### PR DESCRIPTION
**Pull request by jdehaan**
In many solutions people don't implement exactly the described algorithm. They just generate a random number - uniformly - between 3 and 18 to make the checks pass. Some implementations I saw even missed to generate some of the numbers in the range due to an off by one. This test checks that the distribution fits with a +/- 20% tolerance vs the ideal values.

### CODE REVIEW

**Comment by jdehaan** [comment]
If you want to know how I figured out the ideal numbers, as I am lame at probability I just generated all possible combinations and counted the hits with this code:

```csharp
        var combinations = new List<int[]>(6*6*6*6);
        for (int d1 = 1 ; d1 <= 6; d1++)
            for (int d2 = 1 ; d2 <= 6; d2++)
                for (int d3 = 1 ; d3 <= 6; d3++)
                    for (int d4 = 1 ; d4 <= 6; d4++)
                        combinations.Add(new int[] {d1,d2,d3,d4});
        System.Console.WriteLine($"# combinations = {combinations.Count}");
        var stats = combinations
            .GroupBy(x => x.Sum() - x.Min())
            .ToDictionary(g => g.Key, g => g.Count());
        
        foreach(var k in stats.Keys.OrderBy(x => x)) {
            System.Console.WriteLine($"{k,2}: {stats[k],4} -> {stats[k],4}/{6*6*6*6} = {stats[k]*100.0/(6*6*6*6)}");
        }
```

**Comment by jdehaan** [comment]
There is an inherent problem with the test due to the randomness. There could be false negatives... A proper choice of the tolerance and number of runs should be safe enough but there is no guarantee. This is challenging but how else test the correct effect of the implementation? Not testing the distribution leaves too much room for taking some shortcuts and having this test leads people to think about important differences in the implementation, in dnd games this matters...

File: exercises/dnd-character/DndCharacterTest.cs [hunk 1]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
         }
     }
-}
\ No newline at end of file
+    
+    [Fact(Skip = "Remove to run test")]
+    public void Random_ability_is_distributed_correctly()
+    {
+        var idealDistribution = new Dictionary<int, int>() {
+            [3] = 1,            [4] = 4,
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: Any reason for the large number of spaces between the elements? Personally, I think having a little less spaces might be a bit easier to read:

```suggestion
            [3] = 1,        [4] = 4,
```

File: exercises/dnd-character/DndCharacterTest.cs [hunk 2]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
             Assert.Equal(sut.Charisma, sut.Charisma);
         }
     }
-}
\ No newline at end of file
+    
+    [Fact(Skip = "Remove to run test")]
+    public void Random_ability_is_distributed_correctly()
+    {
+        var idealDistribution = new Dictionary<int, int>() {
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: Should it be `idealDistribution` or `expectedDistribution`?

File: exercises/dnd-character/DndCharacterTest.cs [hunk 3]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+        var distribution = new Dictionary<int, int>();
+        int times = 100;
+        for (var i = 3; i <= 18; i++)
+            distribution[i] = 0;
+        for (var i = 0; i < times * 6*6*6*6; i++)
+        {
+            var a = DndCharacter.Ability();
+            distribution[a] += 1;
+        }
+        int min(int ideal) {
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: I think you could make this an expression bodied member : `int min(int ideal) => (int)(ideal * (times * 0.8));`

File: exercises/dnd-character/DndCharacterTest.cs [hunk 4]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+            return (int)(ideal * (times * 0.8));
+        }
+        int max(int ideal) {
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: I think you could make this an expression bodied member : `int max(int ideal) => (int)(ideal * (times * 1.2));`

File: exercises/dnd-character/DndCharacterTest.cs [hunk 5]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+            [5] = 10,           [6] = 21,
+            [7] = 38,           [8] = 62,
+            [9] = 91,           [10] = 122,
+            [11] = 148,         [12] = 167,
+            [13] = 172,         [14] = 160,
+            [15] = 131,         [16] = 94,
+            [17] = 54,          [18] = 21
+        };
+        var distribution = new Dictionary<int, int>();
+        int times = 100;
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: All the other tests files use `var` everywhere:

```suggestion
        var times = 100;
```

File: exercises/dnd-character/DndCharacterTest.cs [hunk 6]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+        for (var i = 3; i <= 18; i++)
+            distribution[i] = 0;
+        for (var i = 0; i < times * 6*6*6*6; i++)
+        {
+            var a = DndCharacter.Ability();
+            distribution[a] += 1;
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]:
```suggestion
            distribution[a]++;
```

File: exercises/dnd-character/DndCharacterTest.cs [hunk 7]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+            [15] = 131,         [16] = 94,
+            [17] = 54,          [18] = 21
+        };
+        var distribution = new Dictionary<int, int>();
+        int times = 100;
+        for (var i = 3; i <= 18; i++)
+            distribution[i] = 0;
+        for (var i = 0; i < times * 6*6*6*6; i++)
+        {
+            var a = DndCharacter.Ability();
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]:
```suggestion
            var score = DndCharacter.Ability();
```

or 

```suggestion
            var ability = DndCharacter.Ability();
```

File: exercises/dnd-character/DndCharacterTest.cs [hunk 8]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+            [3] = 1,            [4] = 4,
+            [5] = 10,           [6] = 21,
+            [7] = 38,           [8] = 62,
+            [9] = 91,           [10] = 122,
+            [11] = 148,         [12] = 167,
+            [13] = 172,         [14] = 160,
+            [15] = 131,         [16] = 94,
+            [17] = 54,          [18] = 21
+        };
+        var distribution = new Dictionary<int, int>();
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: Maybe to contract this with `idealDistribution`, name it:

```suggestion
        var actualDistribution = new Dictionary<int, int>();
```

File: exercises/dnd-character/DndCharacterTest.cs [hunk 9]
```diff
@@ -140,4 +140,36 @@ public void Each_ability_is_only_calculated_once()
...
+        int times = 100;
+        for (var i = 3; i <= 18; i++)
+            distribution[i] = 0;
+        for (var i = 0; i < times * 6*6*6*6; i++)
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: To me, it isn't immediately clear what the significance is of `6*6*6*6`. Perhaps introduce a well-named `const` for it?

File: exercises/dnd-character/DndCharacterTest.cs (hunk 1 above)
Comment (jdehaan) [reply exercises/dnd-character/DndCharacterTest.cs]: Ok

**Comment by jdehaan** [comment]
> You are currently directly editing the test file, however most of the tests (including this one), are automatically generated by our [test generators](https://github.com/exercism/csharp/blob/master/docs/GENERATORS.md). The source for this generator can be found [here](https://github.com/exercism/csharp/blob/master/generators/Exercises/Generators/DndCharacter.cs). Could you incorporate your change in the test generator?

Sure, that makes sense. But I need some time to get into it.

**Comment by ErikSchierboom** [comment]
> Sure, that makes sense. But I need some time to get into it.

Sure! Let us know if you need any help.

**Comment by jdehaan** [comment]
Thx for the review, I quickly cleaned up the current contents. I'll dig into the generator and transfer the code later into it and extend the PR.

File: exercises/dnd-character/DndCharacterTest.cs [hunk 10]
```diff
@@ -140,4 +140,33 @@ public void Each_ability_is_only_calculated_once()
...
+            [7] = 38,       [8] = 62,
+            [9] = 91,       [10] = 122,
+            [11] = 148,     [12] = 167,
+            [13] = 172,     [14] = 160,
+            [15] = 131,     [16] = 94,
+            [17] = 54,      [18] = 21
+        };
+        var actualDistribution = new Dictionary<int, int>();
+        var times = 100;
+        const int POSSIBLE_COMBINATIONS_COUNT = 6*6*6*6; // 4d6
```
Comment (ErikSchierboom) [inline exercises/dnd-character/DndCharacterTest.cs]: Tiny nit:

```suggestion
        const int PossibleCombinationsCount = 6*6*6*6; // 4d6
```

**Comment by jdehaan** [comment]
@ErikSchierboom I had a short look at the generators (I need to dig more into that) and saw that it might be easier to generate a set of assertions with the values calculated within the genrator. What is your opinion, should the resulting test be readable (like it is now) or might it be a bit cryptic? I would opt for the first option (but will make the generator "ugly" in the sense it'll contain much hard coded text)

I think this might be common trade-off between maintainability of the generator vs readability of the result... and therefore I am looking forward to getting some stronger opinion.

**Comment by ErikSchierboom** [comment]
Definitely go for the readable test. Don't worry about the generator being less pretty or having hardcoded strings, this is an exceptional case after all.

**Comment by jdehaan** [comment]
*I consider the contribution as ready*

@ErikSchierboom that was not too difficult :-) if the way I did it with a verbatim data addition to the generator.

I had some trouble running all the generators under Linux (Ubuntu 18.04), installed libcurl3 and could only run generators partly. I re-generated the content using `dotnet run -e DndCharacter` that fortunately worked well and produced the result I committed.

---

BTW: I tried to figure out how to contribute something to the canonical data but that kind of test does not yet seem to be part of the "DSL", the semantics used are constrained and I thing this would end up in testing magic ranges somehow... That is a challenging task to level up the canonical description, but would be a good thing for all language tracks.

**Comment by ErikSchierboom** [comment]
Great work! 

> That is a challenging task to level up the canonical description, but would be a good thing for all language tracks.

Yeah, I agree. This is hard to do unfortunately.

**Pull request merged**